File: dart/ant_ffi/test/key_derivation_test.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:test/test.dart';
import 'package:ant_ffi/ant_ffi.dart';

void main() {
  group('DerivationIndex', () {
    test('random creates valid 32-byte index', () {
      final index = DerivationIndex.random();
      expect(index.toBytes().length, equals(32));
      index.dispose();
    });

    test('random creates unique indices', () {
      final index1 = DerivationIndex.random();
      final index2 = DerivationIndex.random();

      expect(index1.toBytes(), isNot(equals(index2.toBytes())));

      index1.dispose();
      index2.dispose();
    });

    test('fromBytes roundtrip', () {
      final index1 = DerivationIndex.random();
      final bytes = index1.toBytes();

      final index2 = DerivationIndex.fromBytes(bytes);
      expect(index2.toBytes(), equals(bytes));

      index1.dispose();
      index2.dispose();
    });

    test('fromBytes rejects wrong length', () {
      expect(
        () => DerivationIndex.fromBytes(Uint8List(16)),
        throwsA(isA<ArgumentError>()),
      );
    });

    test('throws after dispose', () {
      final index = DerivationIndex.random();
      index.dispose();

      expect(() => index.toBytes(), throwsStateError);
    });
  });

  group('Signature', () {
    test('toBytes returns 96 bytes', () {
      final key = MainSecretKey.random();
      final message = utf8.encode('Test message');
      final signature = key.sign(Uint8List.fromList(message));

      expect(signature.toBytes().length, equals(96));

      signature.dispose();
      key.dispose();
    });

    test('fromBytes roundtrip', () {
      final key = MainSecretKey.random();
      final message = utf8.encode('Test message');
      final sig1 = key.sign(Uint8List.fromList(message));
      final bytes = sig1.toBytes();

      final sig2 = Signature.fromBytes(bytes);
      expect(sig2.toBytes(), equals(bytes));

      sig1.dispose();
      sig2.dispose();
      key.dispose();
    });

    test('fromBytes rejects wrong length', () {
      expect(
        () => Signature.fromBytes(Uint8List(32)),
        throwsA(isA<ArgumentError>()),
      );
    });

    test('toHex returns 192 characters', () {
      final key = MainSecretKey.random();
      final message = utf8.encode('Test message');
      final signature = key.sign(Uint8List.fromList(message));

      final hex = signature.toHex();
      expect(hex.length, equals(192));
      expect(hex, matches(RegExp(r'^[0-9a-f]+$', caseSensitive: false)));

      signature.dispose();
      key.dispose();
    });

    test('parity returns boolean', () {
      final key = MainSecretKey.random();
      final signature = key.sign(Uint8List.fromList(utf8.encode('Test')));

      // Just verify it returns a valid boolean without throwing
      expect(signature.parity(), isA<bool>());

      signature.dispose();
      key.dispose();
    });

    test('throws after dispose', () {
      final key = MainSecretKey.random();
      final signature = key.sign(Uint8List.fromList(utf8.encode('Test')));
      signature.dispose();
      key.dispose();

      expect(() => signature.toBytes(), throwsStateError);
    });
  });

  group('MainSecretKey', () {
    test('random creates valid key', () {
      final key = MainSecretKey.random();
      expect(key.toBytes(), isNotEmpty);
      key.dispose();
    });

    test('random creates unique keys', () {
      final key1 = MainSecretKey.random();
      final key2 = MainSecretKey.random();

      expect(key1.toBytes(), isNot(equals(key2.toBytes())));

      key1.dispose();
      key2.dispose();
    });

    test('fromSecretKey creates valid key', () {
      final secretKey = SecretKey.random();
      final mainKey = MainSecretKey.fromSecretKey(secretKey);

      expect(mainKey.toBytes(), isNotEmpty);

      mainKey.dispose();
      secretKey.dispose();
    });

    test('publicKey returns valid MainPubkey', () {
      final secretKey = MainSecretKey.random();
      final publicKey = secretKey.publicKey();

      expect(publicKey.toHex(), isNotEmpty);

      publicKey.dispose();
      secretKey.dispose();
    });

    test('sign and verify', () {
      final secretKey = MainSecretKey.random();
      final publicKey = secretKey.publicKey();
      final message = Uint8List.fromList(utf8.encode('Hello, world!'));

      final signature = secretKey.sign(message);

      expect(publicKey.verify(signature, message), isTrue);

      signature.dispose();
      publicKey.dispose();
      secretKey.dispose();
    });

    test('signature invalid for wrong message', () {
      final secretKey = MainSecretKey.random();
      final publicKey = secretKey.publicKey();

      final signature = secretKey.sign(Uint8List.fromList(utf8.encode('correct message')));

      expect(
        publicKey.verify(signature, Uint8List.fromList(utf8.encode('wrong message'))),
        isFalse,
      );

      signature.dispose();
      publicKey.dispose();
      secretKey.dispose();
    });

    test('deriveKey returns DerivedSecretKey', () {
      final mainKey = MainSecretKey.random();
      final index = DerivationIndex.random();

      final derivedKey = mainKey.deriveKey(index);

      expect(derivedKey.publicKey().toHex(), isNotEmpty);

      derivedKey.dispose();
      index.dispose();
      mainKey.dispose();
    });

    test('randomDerivedKey returns DerivedSecretKey', () {
      final mainKey = MainSecretKey.random();

      final derivedKey = mainKey.randomDerivedKey();

      expect(derivedKey.publicKey().toHex(), isNotEmpty);

      derivedKey.dispose();
      mainKey.dispose();
    });

    test('multiple randomDerivedKey calls return different keys', () {
      final mainKey = MainSecretKey.random();

      final derived1 = mainKey.randomDerivedKey();
      final derived2 = mainKey.randomDerivedKey();

      expect(derived1.publicKey().toHex(), isNot(equals(derived2.publicKey().toHex())));

      derived1.dispose();
      derived2.dispose();
      mainKey.dispose();
    });

    test('throws after dispose', () {
      final key = MainSecretKey.random();
      key.dispose();

      expect(() => key.toBytes(), throwsStateError);
    });
  });

  group('MainPubkey', () {
    test('fromPublicKey creates valid key', () {
      final secretKey = SecretKey.random();
      final publicKey = secretKey.publicKey();

      final mainPubkey = MainPubkey.fromPublicKey(publicKey);

      expect(mainPubkey.toHex(), isNotEmpty);

      mainPubkey.dispose();
      publicKey.dispose();
      secretKey.dispose();
    });

    test('fromHex roundtrip', () {
      final mainKey = MainSecretKey.random();
      final pubkey = mainKey.publicKey();
      final hex = pubkey.toHex();

      final pubkey2 = MainPubkey.fromHex(hex);
      expect(pubkey2.toHex(), equals(hex));

      pubkey2.dispose();
      pubkey.dispose();
      mainKey.dispose();
    });

    test('toBytes returns valid bytes', () {
      final mainKey = MainSecretKey.random();
      final pubkey = mainKey.publicKey();

      expect(pubkey.toBytes(), isNotEmpty);

      pubkey.dispose();
      mainKey.dispose();
    });

    test('deriveKey returns DerivedPubkey', () {
      final mainKey = MainSecretKey.random();
      final mainPubkey = mainKey.publicKey();
      final index = DerivationIndex.random();

      final derivedPubkey = mainPubkey.deriveKey(index);

      expect(derivedPubkey.toHex(), isNotEmpty);

      derivedPubkey.dispose();
      index.dispose();
      mainPubkey.dispose();
      mainKey.dispose();
    });

    test('throws after dispose', () {
      final key = MainSecretKey.random();
      final pubkey = key.publicKey();
      pubkey.dispose();
      key.dispose();

      expect(() => pubkey.toHex(), throwsStateError);
    });
  });

  group('Key Derivation Consistency', () {
    test('derived keys match when using same index', () {
      final mainSecretKey = MainSecretKey.random();
      final mainPubkey = mainSecretKey.publicKey();
      final index = DerivationIndex.random();

      final derivedSecretKey = mainSecretKey.deriveKey(index);
      final derivedPubkeyFromSecret = derivedSecretKey.publicKey();

      final derivedPubkeyFromMain = mainPubkey.deriveKey(index);

      expect(
        derivedPubkeyFromSecret.toHex(),
        equals(derivedPubkeyFromMain.toHex()),
      );

      derivedPubkeyFromSecret.dispose();
      derivedPubkeyFromMain.dispose();
      derivedSecretKey.dispose();
      index.dispose();
      mainPubkey.dispose();
      mainSecretKey.dispose();
    });
  });

  group('DerivedSecretKey', () {
    test('fromSecretKey creates valid key', () {
      final secretKey = SecretKey.random();
      final derivedKey = DerivedSecretKey.fromSecretKey(secretKey);

      expect(derivedKey.publicKey().toHex(), isNotEmpty);

      derivedKey.dispose();
      secretKey.dispose();
    });

    test('publicKey returns DerivedPubkey', () {
      final mainKey = MainSecretKey.random();
      final derivedKey = mainKey.randomDerivedKey();

      final publicKey = derivedKey.publicKey();

      expect(publicKey.toHex(), isNotEmpty);

      publicKey.dispose();
      derivedKey.dispose();
      mainKey.dispose();
    });

    test('sign and verify', () {
      final mainKey = MainSecretKey.random();
      final derivedKey = mainKey.randomDerivedKey();
      final derivedPubkey = derivedKey.publicKey();
      final message = Uint8List.fromList(utf8.encode('Test message for derived key'));

      final signature = derivedKey.sign(message);

      expect(derivedPubkey.verify(signature, message), isTrue);

      signature.dispose();
      derivedPubkey.dispose();
      derivedKey.dispose();
      mainKey.dispose();
    });

    test('throws after dispose', () {
      final mainKey = MainSecretKey.random();
      final derivedKey = mainKey.randomDerivedKey();
      derivedKey.dispose();
      mainKey.dispose();

      expect(() => derivedKey.publicKey(), throwsStateError);
    });
  });

  group('DerivedPubkey', () {
    test('fromPublicKey creates valid key', () {
      final secretKey = SecretKey.random();
      final publicKey = secretKey.publicKey();

      final derivedPubkey = DerivedPubkey.fromPublicKey(publicKey);

      expect(derivedPubkey.toHex(), isNotEmpty);

      derivedPubkey.dispose();
      publicKey.dispose();
      secretKey.dispose();
    });

    test('fromHex roundtrip', () {
      final mainKey = MainSecretKey.random();
      final derivedKey = mainKey.randomDerivedKey();
      final pubkey = derivedKey.publicKey();
      final hex = pubkey.toHex();

      final pubkey2 = DerivedPubkey.fromHex(hex);
      expect(pubkey2.toHex(), equals(hex));

      pubkey2.dispose();
      pubkey.dispose();
      derivedKey.dispose();
      mainKey.dispose();
    });

    test('toBytes returns valid bytes', () {
      final mainKey = MainSecretKey.random();
      final derivedKey = mainKey.randomDerivedKey();
      final pubkey = derivedKey.publicKey();

      expect(pubkey.toBytes(), isNotEmpty);

      pubkey.dispose();
      derivedKey.dispose();
      mainKey.dispose();
    });

    test('throws after dispose', () {
      final mainKey = MainSecretKey.random();
      final derivedKey = mainKey.randomDerivedKey();
      final pubkey = derivedKey.publicKey();
      pubkey.dispose();
      derivedKey.dispose();
      mainKey.dispose();

      expect(() => pubkey.toHex(), throwsStateError);
    });
  });
}

File: php/ant_ffi/tests/KeyDerivationTest.php
<?php

declare(strict_types=1);

namespace AntFfi\Tests;

use PHPUnit\Framework\TestCase;
use AntFfi\AntFfiException;
use AntFfi\Types\DerivationIndex;
use AntFfi\Types\Signature;
use AntFfi\Types\MainSecretKey;
use AntFfi\Types\MainPubkey;
use AntFfi\Types\DerivedSecretKey;
use AntFfi\Types\DerivedPubkey;
use AntFfi\Types\SecretKey;
use AntFfi\Types\PublicKey;

final class KeyDerivationTest extends TestCase
{
    // DerivationIndex tests

    public function testDerivationIndexRandom(): void
    {
        $index = DerivationIndex::random();
        $bytes = $index->toBytes();

        $this->assertEquals(32, strlen($bytes), 'DerivationIndex should be 32 bytes');
    }

    public function testDerivationIndexFromBytesRoundtrip(): void
    {
        $index1 = DerivationIndex::random();
        $bytes = $index1->toBytes();

        $index2 = DerivationIndex::fromBytes($bytes);
        $this->assertEquals($bytes, $index2->toBytes());
    }

    public function testDerivationIndexFromBytesRejectsWrongLength(): void
    {
        $this->expectException(AntFfiException::class);
        $this->expectExceptionMessage('DerivationIndex must be exactly 32 bytes');
        DerivationIndex::fromBytes('short');
    }

    public function testMultipleRandomIndicesAreDifferent(): void
    {
        $index1 = DerivationIndex::random();
        $index2 = DerivationIndex::random();

        $this->assertNotEquals($index1->toBytes(), $index2->toBytes());
    }

    // MainSecretKey tests

    public function testMainSecretKeyRandom(): void
    {
        $key = MainSecretKey::random();
        $bytes = $key->toBytes();

        $this->assertNotEmpty($bytes);
    }

    public function testMainSecretKeyFromSecretKey(): void
    {
        $secretKey = SecretKey::random();
        $mainKey = MainSecretKey::fromSecretKey($secretKey);

        $this->assertNotEmpty($mainKey->toBytes());
    }

    public function testMainSecretKeyPublicKey(): void
    {
        $secretKey = MainSecretKey::random();
        $publicKey = $secretKey->publicKey();

        $this->assertNotEmpty($publicKey->toHex());
    }

    public function testMainSecretKeySignAndVerify(): void
    {
        $secretKey = MainSecretKey::random();
        $publicKey = $secretKey->publicKey();
        $message = 'Hello, world!';

        $signature = $secretKey->sign($message);

        $this->assertTrue($publicKey->verify($signature, $message));
    }

    public function testMainSecretKeySignatureInvalidForWrongMessage(): void
    {
        $secretKey = MainSecretKey::random();
        $publicKey = $secretKey->publicKey();

        $signature = $secretKey->sign('correct message');

        $this->assertFalse($publicKey->verify($signature, 'wrong message'));
    }

    public function testMainSecretKeyDeriveKey(): void
    {
        $mainKey = MainSecretKey::random();
        $index = DerivationIndex::random();

        $derivedKey = $mainKey->deriveKey($index);

        $this->assertInstanceOf(DerivedSecretKey::class, $derivedKey);
    }

    public function testMainSecretKeyRandomDerivedKey(): void
    {
        $mainKey = MainSecretKey::random();

        $derivedKey = $mainKey->randomDerivedKey();

        $this->assertInstanceOf(DerivedSecretKey::class, $derivedKey);
    }

    public function testMultipleRandomDerivedKeysAreDifferent(): void
    {
        $mainKey = MainSecretKey::random();

        $derived1 = $mainKey->randomDerivedKey();
        $derived2 = $mainKey->randomDerivedKey();

        $this->assertNotEquals(
            $derived1->publicKey()->toHex(),
            $derived2->publicKey()->toHex()
        );
    }

    // MainPubkey tests

    public function testMainPubkeyFromPublicKey(): void
    {
        $secretKey = SecretKey::random();
        $publicKey = $secretKey->publicKey();

        $mainPubkey = MainPubkey::fromPublicKey($publicKey);

        $this->assertNotEmpty($mainPubkey->toHex());
    }

    public function testMainPubkeyFromHexRoundtrip(): void
    {
        $mainKey = MainSecretKey::random();
        $pubkey = $mainKey->publicKey();
        $hex = $pubkey->toHex();

        $pubkey2 = MainPubkey::fromHex($hex);
        $this->assertEquals($hex, $pubkey2->toHex());
    }

    public function testMainPubkeyToBytesRoundtrip(): void
    {
        $mainKey = MainSecretKey::random();
        $pubkey = $mainKey->publicKey();
        $bytes = $pubkey->toBytes();

        $this->assertNotEmpty($bytes);
    }

    public function testMainPubkeyDeriveKey(): void
    {
        $mainKey = MainSecretKey::random();
        $mainPubkey = $mainKey->publicKey();
        $index = DerivationIndex::random();

        $derivedPubkey = $mainPubkey->deriveKey($index);

        $this->assertInstanceOf(DerivedPubkey::class, $derivedPubkey);
    }

    public function testDerivedKeyConsistency(): void
    {
        // Derive secret and public keys with the same index
        // and verify they match
        $mainSecretKey = MainSecretKey::random();
        $mainPubkey = $mainSecretKey->publicKey();
        $index = DerivationIndex::random();

        $derivedSecretKey = $mainSecretKey->deriveKey($index);
        $derivedPubkeyFromSecret = $derivedSecretKey->publicKey();

        $derivedPubkeyFromMain = $mainPubkey->deriveKey($index);

        $this->assertEquals(
            $derivedPubkeyFromSecret->toHex(),
            $derivedPubkeyFromMain->toHex(),
            'Derived public keys should match whether derived from secret or public main key'
        );
    }

    // DerivedSecretKey tests

    public function testDerivedSecretKeyFromSecretKey(): void
    {
        $secretKey = SecretKey::random();
        $derivedKey = DerivedSecretKey::fromSecretKey($secretKey);

        $this->assertInstanceOf(DerivedSecretKey::class, $derivedKey);
    }

    public function testDerivedSecretKeyPublicKey(): void
    {
        $mainKey = MainSecretKey::random();
        $derivedKey = $mainKey->randomDerivedKey();

        $publicKey = $derivedKey->publicKey();

        $this->assertNotEmpty($publicKey->toHex());
    }

    public function testDerivedSecretKeySignAndVerify(): void
    {
        $mainKey = MainSecretKey::random();
        $derivedKey = $mainKey->randomDerivedKey();
        $derivedPubkey = $derivedKey->publicKey();
        $message = 'Test message for derived key';

        $signature = $derivedKey->sign($message);

        $this->assertTrue($derivedPubkey->verify($signature, $message));
    }

    // DerivedPubkey tests

    public function testDerivedPubkeyFromPublicKey(): void
    {
        $secretKey = SecretKey::random();
        $publicKey = $secretKey->publicKey();

        $derivedPubkey = DerivedPubkey::fromPublicKey($publicKey);

        $this->assertNotEmpty($derivedPubkey->toHex());
    }

    public function testDerivedPubkeyFromHexRoundtrip(): void
    {
        $mainKey = MainSecretKey::random();
        $derivedKey = $mainKey->randomDerivedKey();
        $pubkey = $derivedKey->publicKey();
        $hex = $pubkey->toHex();

        $pubkey2 = DerivedPubkey::fromHex($hex);
        $this->assertEquals($hex, $pubkey2->toHex());
    }

    public function testDerivedPubkeyToBytesRoundtrip(): void
    {
        $mainKey = MainSecretKey::random();
        $derivedKey = $mainKey->randomDerivedKey();
        $pubkey = $derivedKey->publicKey();
        $bytes = $pubkey->toBytes();

        $this->assertNotEmpty($bytes);
    }

    // Signature tests

    public function testSignatureToBytes(): void
    {
        $mainKey = MainSecretKey::random();
        $signature = $mainKey->sign('Test message');

        $bytes = $signature->toBytes();
        $this->assertEquals(96, strlen($bytes), 'BLS signature should be 96 bytes');
    }

    public function testSignatureFromBytesRoundtrip(): void
    {
        $mainKey = MainSecretKey::random();
        $sig1 = $mainKey->sign('Test message');
        $bytes = $sig1->toBytes();

        $sig2 = Signature::fromBytes($bytes);
        $this->assertEquals($bytes, $sig2->toBytes());
    }

    public function testSignatureFromBytesRejectsWrongLength(): void
    {
        $this->expectException(AntFfiException::class);
        $this->expectExceptionMessage('Signature must be exactly 96 bytes');
        Signature::fromBytes('short');
    }

    public function testSignatureToHex(): void
    {
        $mainKey = MainSecretKey::random();
        $signature = $mainKey->sign('Test message');

        $hex = $signature->toHex();
        $this->assertMatchesRegularExpression('/^[0-9a-f]+$/i', $hex);
        $this->assertEquals(192, strlen($hex), 'Hex signature should be 192 characters (96 bytes * 2)');
    }

    public function testSignatureParity(): void
    {
        $mainKey = MainSecretKey::random();
        $signature = $mainKey->sign('Test message');

        // Parity returns a boolean - just verify it doesn't throw
        $parity = $signature->parity();
        $this->assertIsBool($parity);
    }

    // Disposal tests

    public function testKeyDisposal(): void
    {
        $key = MainSecretKey::random();
        $this->assertFalse($key->isDisposed());

        $key->dispose();
        $this->assertTrue($key->isDisposed());
    }

    public function testDerivedKeyDisposal(): void
    {
        $mainKey = MainSecretKey::random();
        $derivedKey = $mainKey->randomDerivedKey();
        $this->assertFalse($derivedKey->isDisposed());

        $derivedKey->dispose();
        $this->assertTrue($derivedKey->isDisposed());
    }

    public function testDerivationIndexDisposal(): void
    {
        $index = DerivationIndex::random();
        $this->assertFalse($index->isDisposed());

        $index->dispose();
        $this->assertTrue($index->isDisposed());
    }
}
